File: src/java_version/arrays/CountingSort.java
package java_version.arrays;

import static java_version.arrays.Utils.findMaximumOfArray;
import static java_version.arrays.Utils.findMinimumOfarray;

public class CountingSort {

    public static void main(String[] args) {
        int[] arr = {5, 6 ,7 , 324 ,123 , 123 , 5151 ,51 , 1 , 123 ,15};
        long startTime = System.nanoTime();
        int minVal = findMinimumOfarray(arr);
        int max = findMaximumOfArray(arr);
        countingSort(arr, minVal, max);
        System.out.print("Sorted Ended: " + (System.nanoTime() - startTime) + "\n");
        for (int i : arr) {
            System.out.print("" + i + "\n");
        }
    }

    static void countingSort(int[] array, int minVal, int maxVal) {
        int[] countArray = new int[maxVal - minVal + 1];

        for (int value : array) {
            countArray[value - minVal]++;
        }

        int j = 0;
        for (int i = minVal; i <= maxVal; i++) {
            while (countArray[i - minVal] > 0) {
                array[j++] = i;
                countArray[i - minVal]--;
            }
        }
    }
}

File: src/kotlin_version/arrays/CountingSort.kt
package kotlin_version.arrays

fun main() {
    val startTime = System.nanoTime()

    val arrayOfInts = intArrayOf(1, 3, 45, 151, 13, 1, 5, 2, 4, 5, 6, 7, 8, 10)

    countingSort(arrayOfInts, arrayOfInts.min()!!, arrayOfInts.max()!!)

    val endTime = System.nanoTime()

    println("Execution time: ${endTime - startTime}")

    arrayOfInts.forEach {
        println(it)
    }
}

fun countingSort(array: IntArray, minVal: Int, maxVal: Int) {
    val countingArray = IntArray(maxVal - minVal + 1)

    array.forEach {
        countingArray[it - minVal]++
    }

    var j = 0

    for (i in minVal..maxVal) {
        while (countingArray[i - minVal] > 0) {
            array[j++] = i
            countingArray[i - minVal]--
        }
    }
}